File: src/hierarchical_ot/lp_solvers/HPR-LP-C/bindings/matlab/examples/example_direct_lp.m
% Example: Solving a simple LP problem directly with HPRLP
%
% This example demonstrates how to solve a linear programming problem
% by directly providing the problem data using the Model-based API.
%
% Problem:
%     minimize    -3*x1 - 5*x2
%     subject to   x1 + 2*x2 <= 10
%                 3*x1 +  x2 <= 12
%                  x1, x2 >= 0
%
% Expected solution: x1 ≈ 2.8, x2 ≈ 3.6, objective ≈ -26.4

clear; clc;

% Add HPRLP to path
script_dir = fileparts(mfilename('fullpath'));
addpath(fullfile(script_dir, '..'));

fprintf('======================================================================\n');
fprintf('HPRLP Example: Direct LP Solution\n');
fprintf('======================================================================\n');
fprintf('\n');

%% Define the LP problem

% Constraint matrix in sparse format
% Row 1: x1 + 2*x2 <= 10
% Row 2: 3*x1 + x2 <= 12
A = sparse([1.0, 2.0;
            3.0, 1.0]);

% Constraint bounds (convert inequalities to double-sided bounds)
AL = [-inf; -inf];  % Lower bounds (no lower constraints)
AU = [10.0; 12.0];  % Upper bounds

% Variable bounds
l = [0.0; 0.0];     % Lower bounds (x >= 0)
u = [inf; inf];     % Upper bounds (unbounded above)

% Objective coefficients (minimize -3*x1 - 5*x2)
c = [-3.0; -5.0];

%% Create model

fprintf('Creating model...\n');
model = hprlp.Model.from_arrays(A, AL, AU, l, u, c);
fprintf('Model created: %d constraints, %d variables\n', model.m, model.n);
fprintf('\n');

%% Solve with parameters

fprintf('Solving with parameters (stop_tol=1e-9)...\n');
fprintf('----------------------------------------------------------------------\n');

param = hprlp.Parameters();
param.device_number = 0;
param.stop_tol = 1e-9;

result = model.solve(param);

%% Display results

fprintf('\n');
fprintf('======================================================================\n');
fprintf('Solution Results\n');
fprintf('======================================================================\n');
fprintf('\n');

fprintf('Status:         %s\n', result.status);
fprintf('Objective:      %.6f\n', result.primal_obj);
fprintf('Solution:       x = [%.6f, %.6f]\n', result.x(1), result.x(2));
fprintf('Iterations:     %d\n', result.iter);
fprintf('Time:           %.4f seconds\n', result.time);
fprintf('Duality gap:    %.6e\n', result.gap);
fprintf('Residuals:      %.6e\n', result.residuals);
fprintf('\n');

%% Verify solution

if strcmp(result.status, 'OPTIMAL')
    fprintf('Verification:\n');
    fprintf('  Constraint 1:  x1 + 2*x2 = %.6f <= 10\n', result.x(1) + 2*result.x(2));
    fprintf('  Constraint 2:  3*x1 + x2 = %.6f <= 12\n', 3*result.x(1) + result.x(2));
    fprintf('  Objective:     -3*x1 - 5*x2 = %.6f\n', -3*result.x(1) - 5*result.x(2));
    fprintf('\n');
    fprintf('✓ Solution is optimal!\n');
else
    fprintf('⚠ Solution is not optimal!\n');
end

fprintf('\n');
fprintf('======================================================================\n');

% Note: Model is automatically freed when it goes out of scope (destructor)

File: src/hierarchical_ot/lp_solvers/HPR-LP-C/bindings/julia/examples/example_direct_lp.jl
#!/usr/bin/env julia

"""
Example: Direct LP Data Entry with HPRLP.jl

This example demonstrates solving an LP problem by directly providing
constraint matrices and vectors using the Model-based API.

Problem from model.mps:
    minimize    -3x₁ - 5x₂
    subject to  x₁ + 2x₂ <= 10
                3x₁ + x₂ <= 12
                x₁, x₂ >= 0

Expected solution: x₁ ≈ 2.8, x₂ ≈ 3.6, obj ≈ -26.4
"""

# Add parent directory to load path
push!(LOAD_PATH, joinpath(@__DIR__, ".."))

using HPRLP
using SparseArrays
using LinearAlgebra

println("=" ^ 70)
println("HPRLP.jl Example: Direct LP Input")
println("=" ^ 70)
println()

println("Problem:")
println("  minimize    -3x₁ - 5x₂")
println("  subject to  x₁ + 2x₂ <= 10")
println("              3x₁ + x₂ <= 12")
println("              x₁, x₂ >= 0")
println()

# Constraint matrix A (sparse format)
# Row 0: x₁ + 2x₂ <= 10
# Row 1: 3x₁ + x₂ <= 12
A = sparse([1.0 2.0;
            3.0 1.0])

println("Constraint matrix A:")
println(Matrix(A))  # Display as dense for readability
println()

# Constraint bounds: AL <= Ax <= AU
AL = [-Inf, -Inf]  # No lower bounds
AU = [10.0, 12.0]   # Upper bounds (RHS values)

# Variable bounds: l <= x <= u
l = [0.0, 0.0]      # Lower bounds (x >= 0)
u = [Inf, Inf]      # No upper bounds

# Objective: minimize c'x
c = [-3.0, -5.0]

# Create model
println("Creating model...")
model = Model(A, AL, AU, l, u, c)
println("Model created: $(model.m) constraints, $(model.n) variables")
println()

# Configure solver parameters
println("Configuring solver parameters...")
params = Parameters(
    stop_tol = 1e-9,
    device_number = 0
)
println()

# Solve with parameters
println("Solving with parameters (stop_tol=1e-9)...")
println()

result = solve(model, params)

# Display results
println("=" ^ 70)
println("Solution Results")
println("=" ^ 70)
println()

println("Status:         ", result.status)
println("Objective:      ", round(result.primal_obj, digits=6))
println("Solution:       x = [", join(round.(result.x, digits=6), ", "), "]")
println("Iterations:     ", result.iter)
println("Time:           ", round(result.time, digits=4), " seconds")
println("Duality gap:    ", result.gap)
println("Residuals:      ", result.residuals)
println()

# Verify solution
if result.status == "OPTIMAL"
    x = result.x
    println("Verification:")
    println("  Constraint 1:  x₁ + 2x₂ = ", round(x[1] + 2*x[2], digits=6), " <= 10")
    println("  Constraint 2:  3x₁ + x₂ = ", round(3*x[1] + x[2], digits=6), " <= 12")
    println("  Objective:     -3x₁ - 5x₂ = ", round(-3*x[1] - 5*x[2], digits=6))
    println()
    println("✓ Solution is optimal!")
else
    println("⚠ Solution is not optimal!")
end

# Note: Model is automatically freed when garbage collected
# Or explicitly: free(model)

println()
println("=" ^ 70)
